File: frontend/src/utils/dayRegionGrouping.ts
import type { AlbumFile } from "@/types";

/**
 * "By day and region" grouping — one section per calendar day, and inside it one sub-section
 * per place, where a place is a set of photos that all sit within `radiusMeters` of each other.
 *
 * The clustering is deliberately *not* a plain "grow a blob while the next photo is near the
 * last one" pass: that chains, so a walk across a city merges into one 20 km "region" even
 * though its ends are nowhere near each other. What runs here is a complete-linkage
 * agglomerative clustering with the merge threshold as a hard cap: two clusters merge only when
 * *every* cross pair is within the radius, so a finished cluster is guaranteed to have a
 * diameter of at most `radiusMeters`. Merges are attempted shortest-edge-first, which is what
 * makes the greedy pass produce the same natural groups a full O(n³) run would.
 */

/** Default region size: photos in one cluster are never more than this far apart. */
export const DEFAULT_REGION_RADIUS_METERS = 2000;

/**
 * Above this many located photos in a single day the exact pass is skipped for a leader pass
 * (see `leaderCluster`). Exact merging is O(n²) in both candidate pairs and cross-pair checks,
 * which is nothing for a normal day out and a frozen tab for a bulk import of a whole year
 * stamped with one date.
 */
const EXACT_PASS_LIMIT = 1500;

const EARTH_RADIUS_METERS = 6_371_008.8;

export interface LatLng {
  lat: number;
  lng: number;
}

export interface RegionCluster {
  /** Stable key for `v-for`, derived from the day and the cluster's first file. */
  key: string;
  files: AlbumFile[];
  /** Mean position of the cluster's photos, or null for the "no location" bucket. */
  center: LatLng | null;
  /** Largest distance between any two photos in the cluster, in metres (0 for a single photo). */
  spreadMeters: number;
  /** False for the one bucket per day that collects photos without GPS coordinates. */
  located: boolean;
}

export interface DayGroup {
  /**
   * Calendar day at the capture location, `YYYY-MM-DD`; `"unknown"` when a photo carries no
   * usable date. Not the viewer's day: see `dayOf`.
   */
  key: string;
  /** That day at local midnight (for labelling only), or null for the unknown-date group. */
  date: Date | null;
  clusters: RegionCluster[];
  count: number;
}

/** Great-circle distance in metres. */
export function haversineMeters(a: LatLng, b: LatLng): number {
  const toRad = Math.PI / 180;
  const dLat = (b.lat - a.lat) * toRad;
  const dLng = (b.lng - a.lng) * toRad;
  const lat1 = a.lat * toRad;
  const lat2 = b.lat * toRad;
  const h =
    Math.sin(dLat / 2) ** 2 + Math.cos(lat1) * Math.cos(lat2) * Math.sin(dLng / 2) ** 2;
  return 2 * EARTH_RADIUS_METERS * Math.asin(Math.min(1, Math.sqrt(h)));
}

/** The capture position of a file, or null when it carries no coordinates. */
export function fileLatLng(file: AlbumFile): LatLng | null {
  if (typeof file.gpsLatitude !== "number" || typeof file.gpsLongitude !== "number") return null;
  if (Number.isNaN(file.gpsLatitude) || Number.isNaN(file.gpsLongitude)) return null;
  return { lat: file.gpsLatitude, lng: file.gpsLongitude };
}

/** When the photo was taken, preferring EXIF over the upload timestamp. A true instant. */
function captureDate(file: AlbumFile): Date | null {
  const raw = file.exifDateTimeOriginal || file.uploadedAt;
  if (!raw) return null;
  const date = new Date(raw);
  return Number.isNaN(date.getTime()) ? null : date;
}

/** The capture offset the server resolved for this file, or null when it never knew one. */
function fileOffsetSeconds(file: AlbumFile): number | null {
  const seconds = file.captureUtcOffsetSeconds;
  return typeof seconds === "number" && Number.isFinite(seconds) ? seconds : null;
}

/**
 * The offset that stands in for files carrying none of their own: the one most of this album's
 * files were shot at.
 *
 * A gallery is normally one trip in one place, so a video with only a zone-less mvhd atom, or a
 * photo whose original was purged before the offset column existed, belongs to the same day
 * boundaries as its neighbours — far closer than the viewer's timezone, which is what the
 * remaining `null` case falls back to.
 */
function dominantOffsetSeconds(files: AlbumFile[]): number | null {
  const tally = new Map<number, number>();
  for (const file of files) {
    const seconds = fileOffsetSeconds(file);
    if (seconds === null) continue;
    tally.set(seconds, (tally.get(seconds) ?? 0) + 1);
  }
  let best: number | null = null;
  let bestCount = 0;
  for (const [seconds, count] of tally) {
    if (count > bestCount) {
      best = seconds;
      bestCount = count;
    }
  }
  return best;
}

/**
 * The calendar day a file belongs to, as `YYYY-MM-DD` plus a Date at that day's local midnight
 * for labelling.
 *
 * Cut on the wall clock where the shutter fired, never on the viewer's: `exifDateTimeOriginal` is
 * a true instant, so reading its day in Frankfurt puts a Toronto photo taken at 19:00 into the
 * next day. Adding the capture offset back and then reading UTC fields gives the camera's own
 * clock. With no offset anywhere (`fallbackOffsetSeconds` null too) the viewer's timezone is all
 * that is left, which is the old behaviour and correct for photos taken at home.
 */
function dayOf(date: Date, offsetSeconds: number | null): { key: string; date: Date } {
  let year: number;
  let month: number;
  let day: number;
  if (offsetSeconds === null) {
    year = date.getFullYear();
    month = date.getMonth();
    day = date.getDate();
  } else {
    const shifted = new Date(date.getTime() + offsetSeconds * 1000);
    year = shifted.getUTCFullYear();
    month = shifted.getUTCMonth();
    day = shifted.getUTCDate();
  }
  const key = `${year}-${String(month + 1).padStart(2, "0")}-${String(day).padStart(2, "0")}`;
  // Local midnight of that day: only its Y/M/D are ever read back, by `formatDayLabel`.
  return { key, date: new Date(year, month, day) };
}

/**
 * Local flat projection in metres around `origin`. Over a few kilometres the error against the
 * ellipsoid is centimetres, and it turns the grid bucketing below into plain arithmetic.
 */
function project(point: LatLng, origin: LatLng): { x: number; y: number } {
  const toRad = Math.PI / 180;
  return {
    x: (point.lng - origin.lng) * toRad * EARTH_RADIUS_METERS * Math.cos(origin.lat * toRad),
    y: (point.lat - origin.lat) * toRad * EARTH_RADIUS_METERS,
  };
}

interface Cluster {
  members: number[];
  spread: number;
}

/**
 * Groups indices of `points` so that no cluster is wider than `radius`.
 *
 * Candidate pairs come from a uniform grid with cells one radius wide, so only the 3×3 cell
 * neighbourhood of each point is examined instead of all pairs. Those pairs are then merged
 * shortest-first under the complete-linkage rule.
 *
 * @returns clusters as arrays of indices into `points`, each with its own diameter
 */
function completeLinkageCluster(points: LatLng[], radius: number): Cluster[] {
  const n = points.length;
  if (n === 0) return [];
  if (n === 1) return [{ members: [0], spread: 0 }];

  const origin = points[0];
  const flat = points.map((p) => project(p, origin));

  // --- Candidate pairs via a radius-wide grid ------------------------------------------------
  const cells = new Map<string, number[]>();
  const cellOf = (i: number) => ({
    cx: Math.floor(flat[i].x / radius),
    cy: Math.floor(flat[i].y / radius),
  });
  for (let i = 0; i < n; i++) {
    const { cx, cy } = cellOf(i);
    const key = `${cx}:${cy}`;
    const bucket = cells.get(key);
    if (bucket) bucket.push(i);
    else cells.set(key, [i]);
  }

  const edges: { a: number; b: number; d: number }[] = [];
  for (let i = 0; i < n; i++) {
    const { cx, cy } = cellOf(i);
    for (let dx = -1; dx <= 1; dx++) {
      for (let dy = -1; dy <= 1; dy++) {
        const bucket = cells.get(`${cx + dx}:${cy + dy}`);
        if (!bucket) continue;
        for (const j of bucket) {
          if (j <= i) continue; // each unordered pair once
          const d = haversineMeters(points[i], points[j]);
          if (d <= radius) edges.push({ a: i, b: j, d });
        }
      }
    }
  }
  edges.sort((p, q) => p.d - q.d);

  // --- Merge shortest-first, but only when the whole merged set stays inside the radius -------
  const owner = new Int32Array(n); // index -> cluster id
  const clusters: (Cluster | null)[] = [];
  for (let i = 0; i < n; i++) {
    owner[i] = i;
    clusters.push({ members: [i], spread: 0 });
  }

  for (const edge of edges) {
    const idA = owner[edge.a];
    const idB = owner[edge.b];
    if (idA === idB) continue;
    const a = clusters[idA]!;
    const b = clusters[idB]!;

    let worst = Math.max(a.spread, b.spread);
    let joinable = true;
    for (const i of a.members) {
      for (const j of b.members) {
        const d = haversineMeters(points[i], points[j]);
        if (d > radius) {
          joinable = false;
          break;
        }
        if (d > worst) worst = d;
      }
      if (!joinable) break;
    }
    if (!joinable) continue;

    // Fold the smaller side in, so the owner rewrite stays cheap.
    const [keepId, dropId] =
      a.members.length >= b.members.length ? [idA, idB] : [idB, idA];
    const keep = clusters[keepId]!;
    const drop = clusters[dropId]!;
    for (const i of drop.members) owner[i] = keepId;
    keep.members.push(...drop.members);
    keep.spread = worst;
    clusters[dropId] = null;
  }

  return clusters.filter((c): c is Cluster => c !== null);
}

/**
 * Fallback for very large days: one pass that opens a new cluster whenever a photo is further
 * than half a radius from every existing cluster's leader. Half, because two photos on opposite
 * sides of a leader are then still within a full radius of each other — the diameter guarantee
 * of the exact pass survives, only the choice of groups is cruder.
 */
function leaderCluster(points: LatLng[], radius: number): Cluster[] {
  const half = radius / 2;
  const leaders: { at: LatLng; cluster: Cluster }[] = [];
  for (let i = 0; i < points.length; i++) {
    let best: { at: LatLng; cluster: Cluster } | null = null;
    let bestDistance = Infinity;
    for (const leader of leaders) {
      const d = haversineMeters(points[i], leader.at);
      if (d <= half && d < bestDistance) {
        best = leader;
        bestDistance = d;
      }
    }
    if (best) {
      best.cluster.members.push(i);
      best.cluster.spread = Math.max(best.cluster.spread, bestDistance * 2);
    } else {
      leaders.push({ at: points[i], cluster: { members: [i], spread: 0 } });
    }
  }
  return leaders.map((l) => l.cluster);
}

function centroid(points: LatLng[]): LatLng {
  // Mean of unit vectors, so a cluster straddling the ±180° meridian still lands in the middle.
  const toRad = Math.PI / 180;
  let x = 0;
  let y = 0;
  let z = 0;
  for (const p of points) {
    const lat = p.lat * toRad;
    const lng = p.lng * toRad;
    x += Math.cos(lat) * Math.cos(lng);
    y += Math.cos(lat) * Math.sin(lng);
    z += Math.sin(lat);
  }
  const n = points.length || 1;
  x /= n;
  y /= n;
  z /= n;
  const hyp = Math.sqrt(x * x + y * y);
  return {
    lat: (Math.atan2(z, hyp) * 180) / Math.PI,
    lng: (Math.atan2(y, x) * 180) / Math.PI,
  };
}

/**
 * Splits files into day sections, each split again into regions of at most `radiusMeters`.
 *
 * Days, the clusters inside a day, and the files inside a cluster all keep the order they had
 * in `files` (the album's own order): the grouping re-shelves the gallery, it does not re-sort
 * it. Photos without a usable date land in a trailing "unknown" day; photos without coordinates
 * land in a trailing bucket of their own day.
 */
export function groupByDayAndRegion(
  files: AlbumFile[],
  radiusMeters: number = DEFAULT_REGION_RADIUS_METERS,
): DayGroup[] {
  const days = new Map<string, { date: Date | null; files: AlbumFile[] }>();
  const fallbackOffsetSeconds = dominantOffsetSeconds(files);

  for (const file of files) {
    const instant = captureDate(file);
    const at = instant
      ? dayOf(instant, fileOffsetSeconds(file) ?? fallbackOffsetSeconds)
      : null;
    const key = at ? at.key : "unknown";
    let day = days.get(key);
    if (!day) {
      day = { date: at ? at.date : null, files: [] };
      days.set(key, day);
    }
    day.files.push(file);
  }

  const groups: DayGroup[] = [];
  for (const [key, day] of days) {
    const located: AlbumFile[] = [];
    const points: LatLng[] = [];
    const unlocated: AlbumFile[] = [];
    for (const file of day.files) {
      const at = fileLatLng(file);
      if (at) {
        located.push(file);
        points.push(at);
      } else {
        unlocated.push(file);
      }
    }

    const raw =
      points.length > EXACT_PASS_LIMIT
        ? leaderCluster(points, radiusMeters)
        : completeLinkageCluster(points, radiusMeters);

    // Album order decides which cluster comes first and how photos sit inside it.
    for (const cluster of raw) cluster.members.sort((a, b) => a - b);
    raw.sort((a, b) => a.members[0] - b.members[0]);

    const clusters: RegionCluster[] = raw.map((cluster) => {
      const clusterFiles = cluster.members.map((i) => located[i]);
      return {
        key: `${key}:loc:${clusterFiles[0].id}`,
        files: clusterFiles,
        center: centroid(cluster.members.map((i) => points[i])),
        spreadMeters: cluster.spread,
        located: true,
      };
    });

    if (unlocated.length > 0) {
      clusters.push({
        key: `${key}:nogps`,
        files: unlocated,
        center: null,
        spreadMeters: 0,
        located: false,
      });
    }

    groups.push({ key, date: day.date, clusters, count: day.files.length });
  }

  return groups;
}

/** "1.2 km" / "800 m" — the span quoted next to a region heading. */
export function formatDistance(meters: number): string {
  if (!Number.isFinite(meters) || meters <= 0) return "0 m";
  if (meters < 1000) return `${Math.round(meters / 10) * 10} m`;
  return `${(meters / 1000).toFixed(1)} km`;
}

/** "48.137, 11.575" — the region label used when no place name is available. */
export function formatCoordinates(at: LatLng): string {
  return `${at.lat.toFixed(3)}, ${at.lng.toFixed(3)}`;
}

/** "Monday, 4 May 2026" in the browser's locale; "Date unknown" for the trailing group. */
export function formatDayLabel(day: DayGroup): string {
  if (!day.date) return "Date unknown";
  return day.date.toLocaleDateString(undefined, {
    weekday: "long",
    year: "numeric",
    month: "long",
    day: "numeric",
  });
}

File: ios/Zyncloud/Utils/DayRegionGrouping.swift
import Foundation

/// "By day and place" — one section per calendar day, and inside it one sub-section per place,
/// where a place is a set of photos that all sit within `radiusMeters` of each other.
///
/// A port of the web app's `dayRegionGrouping.ts`, kept deliberately faithful so the same album
/// breaks into the same sections in both clients. The clustering is *not* a plain "grow a blob
/// while the next photo is near the last one" pass: that chains, so a walk across a city merges
/// into one 20 km "place" even though its ends are nowhere near each other. What runs here is a
/// complete-linkage agglomerative clustering with the merge threshold as a hard cap — two
/// clusters merge only when *every* cross pair is within the radius, so a finished cluster is at
/// most `radiusMeters` across. Merges are tried shortest-edge-first, which is what makes the
/// greedy pass produce the same natural groups a full O(n³) run would.

/// A position in signed decimal degrees (WGS 84), the same frame the server stores.
struct LatLng: Equatable {
    let lat: Double
    let lng: Double
}

/// One place inside a day.
struct RegionCluster: Identifiable {
    let id: String
    let photos: [Photo]

    /// Mean position of the cluster's photos, or nil for the "no location" bucket.
    let center: LatLng?

    /// Largest distance between any two photos in the cluster, in metres (0 for a single one).
    let spreadMeters: Double

    /// False for the one bucket per day that collects photos with no coordinates.
    let located: Bool
}

/// One calendar day.
struct DayGroup: Identifiable {
    /// `YYYY-MM-DD` at the capture location, or `"unknown"` when a photo carries no usable date.
    /// Not the viewer's day — see `dayOf`.
    let id: String

    /// That day at local midnight, for labelling only. Nil for the unknown-date group.
    let date: Date?

    let clusters: [RegionCluster]
    let count: Int
}

/// Default place size: photos in one cluster are never more than this far apart.
let defaultRegionRadiusMeters: Double = 2000

/// Above this many located photos in a single day the exact pass is skipped for a leader pass.
/// Exact merging is O(n²) in both candidate pairs and cross-pair checks, which is nothing for a
/// normal day out and a frozen screen for a bulk import of a whole year stamped with one date.
///
/// Internal rather than private so the tests can build a day either side of it: the two passes
/// are different code, and a test that only ever builds small days never runs the one a real
/// bulk import takes.
let exactPassLimit = 1500

private let earthRadiusMeters: Double = 6_371_008.8

// MARK: - Geometry

/// Great-circle distance in metres.
func haversineMeters(_ a: LatLng, _ b: LatLng) -> Double {
    let toRad = Double.pi / 180
    let dLat = (b.lat - a.lat) * toRad
    let dLng = (b.lng - a.lng) * toRad
    let lat1 = a.lat * toRad
    let lat2 = b.lat * toRad
    let h = pow(sin(dLat / 2), 2) + cos(lat1) * cos(lat2) * pow(sin(dLng / 2), 2)
    return 2 * earthRadiusMeters * asin(min(1, sqrt(h)))
}

/// Local flat projection in metres around `origin`. Over a few kilometres the error against the
/// ellipsoid is centimetres, and it turns the grid bucketing below into plain arithmetic.
private func project(_ point: LatLng, around origin: LatLng) -> (x: Double, y: Double) {
    let toRad = Double.pi / 180
    return (
        x: (point.lng - origin.lng) * toRad * earthRadiusMeters * cos(origin.lat * toRad),
        y: (point.lat - origin.lat) * toRad * earthRadiusMeters
    )
}

private func centroid(of points: [LatLng]) -> LatLng {
    // Mean of unit vectors, so a cluster straddling the ±180° meridian still lands in the middle.
    let toRad = Double.pi / 180
    var x = 0.0, y = 0.0, z = 0.0
    for point in points {
        let lat = point.lat * toRad
        let lng = point.lng * toRad
        x += cos(lat) * cos(lng)
        y += cos(lat) * sin(lng)
        z += sin(lat)
    }
    let n = Double(max(points.count, 1))
    x /= n
    y /= n
    z /= n
    let hyp = (x * x + y * y).squareRoot()
    return LatLng(
        lat: atan2(z, hyp) * 180 / .pi,
        lng: atan2(y, x) * 180 / .pi,
    )
}

// MARK: - Reading a photo

extension Photo {
    /// The capture position, or nil when the photo carries no coordinates.
    var latLng: LatLng? {
        guard let gpsLatitude, let gpsLongitude,
              gpsLatitude.isFinite, gpsLongitude.isFinite
        else {
            return nil
        }
        return LatLng(lat: gpsLatitude, lng: gpsLongitude)
    }

    /// When the photo was taken, preferring what the camera recorded over when it was uploaded.
    /// A true instant, not a wall clock.
    var captureInstant: Date? {
        ISO8601.parse(exifDateTimeOriginal) ?? ISO8601.parse(uploadedAt)
    }
}

/// The server writes instants as ISO-8601, with fractional seconds when it has them and without
/// when it does not, so both shapes have to be tried.
enum ISO8601 {
    private static let withFractionalSeconds: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return formatter
    }()

    private static let plain: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime]
        return formatter
    }()

    static func parse(_ text: String?) -> Date? {
        guard let text, !text.isEmpty else { return nil }
        return withFractionalSeconds.date(from: text) ?? plain.date(from: text)
    }
}

// MARK: - Days

/// The offset that stands in for photos carrying none of their own: the one most of this
/// album's photos were shot at.
///
/// An album is normally one trip in one place, so a video with a zone-less container, or a photo
/// whose original was purged before the offset column existed, belongs to the same day
/// boundaries as its neighbours — far closer than the viewer's timezone, which is what the
/// remaining nil case falls back to.
private func dominantOffsetSeconds(_ photos: [Photo]) -> Int? {
    var tally: [Int: Int] = [:]
    for photo in photos {
        guard let seconds = photo.captureUtcOffsetSeconds else { continue }
        tally[seconds, default: 0] += 1
    }
    // Ties broken by the smaller offset so the answer does not wander between runs.
    return tally.max { left, right in
        left.value != right.value ? left.value < right.value : left.key > right.key
    }?.key
}

/// The calendar day a photo belongs to, as `YYYY-MM-DD` plus a local-midnight `Date` to label it
/// with.
///
/// Cut on the wall clock where the shutter fired, never on the viewer's: the capture instant is a
/// true instant, so reading its day in Frankfurt puts a Toronto photo taken at 19:00 into the
/// next day. Adding the capture offset back and then reading UTC fields gives the camera's own
/// clock. With no offset anywhere the viewer's timezone is all that is left, which is correct for
/// photos taken at home.
func dayOf(_ instant: Date, offsetSeconds: Int?) -> (key: String, date: Date?) {
    var calendar = Calendar(identifier: .gregorian)
    let moment: Date

    if let offsetSeconds {
        calendar.timeZone = TimeZone(secondsFromGMT: 0) ?? .gmt
        moment = instant.addingTimeInterval(TimeInterval(offsetSeconds))
    } else {
        calendar.timeZone = .current
        moment = instant
    }

    let parts = calendar.dateComponents([.year, .month, .day], from: moment)
    guard let year = parts.year, let month = parts.month, let day = parts.day else {
        return ("unknown", nil)
    }

    let key = String(format: "%04d-%02d-%02d", year, month, day)

    // Local midnight of that day: only its year/month/day are ever read back, by the label.
    var localMidnight = Calendar(identifier: .gregorian)
    localMidnight.timeZone = .current
    let date = localMidnight.date(from: DateComponents(year: year, month: month, day: day))

    return (key, date)
}

// MARK: - Clustering

private struct Cluster {
    var members: [Int]
    var spread: Double
}

/// Groups indices of `points` so that no cluster is wider than `radius`.
///
/// Candidate pairs come from a uniform grid with cells one radius wide, so only the 3×3 cell
/// neighbourhood of each point is examined instead of all pairs. Those pairs are then merged
/// shortest-first under the complete-linkage rule.
private func completeLinkageCluster(_ points: [LatLng], radius: Double) -> [Cluster] {
    let n = points.count
    if n == 0 { return [] }
    if n == 1 { return [Cluster(members: [0], spread: 0)] }

    let origin = points[0]
    let flat = points.map { project($0, around: origin) }

    // --- Candidate pairs via a radius-wide grid ---------------------------------------------
    struct Cell: Hashable {
        let x: Int
        let y: Int
    }

    func cell(of index: Int) -> Cell {
        Cell(
            x: Int((flat[index].x / radius).rounded(.down)),
            y: Int((flat[index].y / radius).rounded(.down)),
        )
    }

    var cells: [Cell: [Int]] = [:]
    for index in 0 ..< n {
        cells[cell(of: index), default: []].append(index)
    }

    struct Edge {
        let a: Int
        let b: Int
        let distance: Double
    }

    var edges: [Edge] = []
    for index in 0 ..< n {
        let home = cell(of: index)
        for dx in -1 ... 1 {
            for dy in -1 ... 1 {
                guard let bucket = cells[Cell(x: home.x + dx, y: home.y + dy)] else { continue }
                for other in bucket where other > index { // each unordered pair once
                    let distance = haversineMeters(points[index], points[other])
                    if distance <= radius {
                        edges.append(Edge(a: index, b: other, distance: distance))
                    }
                }
            }
        }
    }
    edges.sort { $0.distance < $1.distance }

    // --- Merge shortest-first, but only when the whole merged set stays inside the radius ----
    var owner = Array(0 ..< n)
    var clusters: [Cluster?] = (0 ..< n).map { Cluster(members: [$0], spread: 0) }

    for edge in edges {
        let idA = owner[edge.a]
        let idB = owner[edge.b]
        if idA == idB { continue }
        guard let a = clusters[idA], let b = clusters[idB] else { continue }

        var worst = max(a.spread, b.spread)
        var joinable = true
        outer: for i in a.members {
            for j in b.members {
                let distance = haversineMeters(points[i], points[j])
                if distance > radius {
                    joinable = false
                    break outer
                }
                if distance > worst { worst = distance }
            }
        }
        if !joinable { continue }

        // Fold the smaller side in, so the owner rewrite stays cheap.
        let keepId = a.members.count >= b.members.count ? idA : idB
        let dropId = keepId == idA ? idB : idA
        guard let drop = clusters[dropId] else { continue }
        for i in drop.members { owner[i] = keepId }
        clusters[keepId]?.members.append(contentsOf: drop.members)
        clusters[keepId]?.spread = worst
        clusters[dropId] = nil
    }

    return clusters.compactMap { $0 }
}

/// Fallback for very large days: one pass that opens a new cluster whenever a photo is further
/// than half a radius from every existing cluster's leader. Half, because two photos on opposite
/// sides of a leader are then still within a full radius of each other — the diameter guarantee
/// of the exact pass survives, only the choice of groups is cruder.
private func leaderCluster(_ points: [LatLng], radius: Double) -> [Cluster] {
    let half = radius / 2
    var leaderPoints: [LatLng] = []
    var clusters: [Cluster] = []

    for index in points.indices {
        var best: Int?
        var bestDistance = Double.infinity
        for leader in leaderPoints.indices {
            let distance = haversineMeters(points[index], leaderPoints[leader])
            if distance <= half, distance < bestDistance {
                best = leader
                bestDistance = distance
            }
        }
        if let best {
            clusters[best].members.append(index)
            clusters[best].spread = max(clusters[best].spread, bestDistance * 2)
        } else {
            leaderPoints.append(points[index])
            clusters.append(Cluster(members: [index], spread: 0))
        }
    }

    return clusters
}

// MARK: - The grouping itself

/// Splits photos into day sections, each split again into places of at most `radiusMeters`.
///
/// Days, the places inside a day, and the photos inside a place all keep the order they had in
/// `photos` (the album's own order): the grouping re-shelves the album, it does not re-sort it.
/// Photos with no usable date land in a trailing "unknown" day; photos with no coordinates land
/// in a trailing bucket of their own day.
func groupByDayAndRegion(
    _ photos: [Photo],
    radiusMeters: Double = defaultRegionRadiusMeters,
) -> [DayGroup] {
    var order: [String] = []
    var days: [String: (date: Date?, photos: [Photo])] = [:]
    let fallbackOffsetSeconds = dominantOffsetSeconds(photos)

    for photo in photos {
        let at = photo.captureInstant.map {
            dayOf($0, offsetSeconds: photo.captureUtcOffsetSeconds ?? fallbackOffsetSeconds)
        }
        let key = at?.key ?? "unknown"
        if days[key] == nil {
            days[key] = (date: at?.date, photos: [])
            order.append(key)
        }
        days[key]?.photos.append(photo)
    }

    return order.compactMap { key -> DayGroup? in
        guard let day = days[key] else { return nil }

        var located: [Photo] = []
        var points: [LatLng] = []
        var unlocated: [Photo] = []
        for photo in day.photos {
            if let at = photo.latLng {
                located.append(photo)
                points.append(at)
            } else {
                unlocated.append(photo)
            }
        }

        var raw = points.count > exactPassLimit
            ? leaderCluster(points, radius: radiusMeters)
            : completeLinkageCluster(points, radius: radiusMeters)

        // Album order decides which place comes first and how photos sit inside it.
        for index in raw.indices {
            raw[index].members.sort()
        }
        raw.sort { ($0.members.first ?? 0) < ($1.members.first ?? 0) }

        var clusters: [RegionCluster] = raw.map { cluster in
            let clusterPhotos = cluster.members.map { located[$0] }
            return RegionCluster(
                id: "\(key):loc:\(clusterPhotos.first?.id ?? 0)",
                photos: clusterPhotos,
                center: centroid(of: cluster.members.map { points[$0] }),
                spreadMeters: cluster.spread,
                located: true,
            )
        }

        if !unlocated.isEmpty {
            clusters.append(RegionCluster(
                id: "\(key):nogps",
                photos: unlocated,
                center: nil,
                spreadMeters: 0,
                located: false,
            ))
        }

        return DayGroup(id: key, date: day.date, clusters: clusters, count: day.photos.count)
    }
}

// MARK: - Labels

/// "1.2 km" / "800 m" — the span quoted next to a place heading.
func formatDistance(_ meters: Double) -> String {
    guard meters.isFinite, meters > 0 else { return "0 m" }
    if meters < 1000 {
        return "\(Int((meters / 10).rounded()) * 10) m"
    }
    return String(format: "%.1f km", meters / 1000)
}

/// "48.137, 11.575" — the place label used when no name is available.
func formatCoordinates(_ at: LatLng) -> String {
    String(format: "%.3f, %.3f", at.lat, at.lng)
}

/// "Monday, 4 May 2026" in the device's locale; "Date unknown" for the trailing group.
func formatDayLabel(_ day: DayGroup) -> String {
    guard let date = day.date else { return "Date unknown" }
    let formatter = DateFormatter()
    formatter.dateStyle = .full
    formatter.timeStyle = .none
    return formatter.string(from: date)
}
